feature: ascore with phospho decoys

timosachsenberg opened the pull request:
## Description

<!-- Please include a summary of the change and which issue is fixed here. -->

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - PhosphoScoring: optional inclusion of decoy phosphorylation sites (A, G, L) for FLR analysis via an "add_decoys" option.
  - AScore now emits a ProFormaString meta value with localization scores.
  - New numerical utilities for binomial/log-sum-exp calculations.

- **Bug Fixes / Stability**
  - AScore rewritten for improved numerical stability and non-recursive performance.

- **Tests**
  - Expanded tests for decoy handling and the new math utilities.

- **Chores**
  - Spectrum preprocessing promoted to the public API.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

timosachsenberg committed:
support PhoshoDecoy

timosachsenberg committed:
decoy phosho annotation

timosachsenberg committed:
add doc

timosachsenberg committed:
Enhance AScore computation to handle PhosphoDecoy events and improve early termination checks for permutations

timosachsenberg committed:
Refactor AScore methods for phospho site detection and update test cases

timosachsenberg committed:
Add mathematical functions for binomial coefficient and log-sum-exp operations with corresponding tests

timosachsenberg committed:
use new methods

timosachsenberg committed:
Refactor AScore to use numerically stable binomial CDF calculation and optimize permutation generation with bitmasking

timosachsenberg committed:
Update CHANGELOG to include PhosphoScoring tool support for PhosphoDecoy and AScore improvements for numerical stability

timosachsenberg commented on src/tests/class_tests/openms/source/MathFunctions_test.cpp:
```diff
@@ -1,174 +1,93 @@
 // --------------------------------------------------------------------------
-// $Maintainer: Timo Sachsenberg$
-// $Authors: Marc Sturm, Chris Bielow $
+// $Maintainer: Benjamin Sachsenhofer $
+// $Authors: Benjamin Sachsenhofer $
```
```suggestion
// $Maintainer: Timo Sachsenberg $
// $Authors: Timo Sachsenberg $
```

timosachsenberg commented on src/tests/class_tests/openms/source/MathFunctions_test.cpp:
```diff
@@ -1,174 +1,93 @@
 // --------------------------------------------------------------------------
-// $Maintainer: Timo Sachsenberg$
-// $Authors: Marc Sturm, Chris Bielow $
+// $Maintainer: Timo Sachsenberg $
+// $Authors: Timo Sachsenberg $
```
```suggestion
// $Maintainer: Timo Sachsenberg $
// $Authors: Marc Sturm, Chris Bielow, Timo Sachsenberg $
```

timosachsenberg committed:
more comments

timosachsenberg committed:
properly configure TSG

jpfeuffer commented:
I think this could be improved a lot by *actually* calculating cumulative scores and not always the full binomial probability.
Then you also would not need the weird permutation generation in the beginning (for which I am not sure if I like the bitmask version better).

But if this fixes things, go ahead. The stable version at least somewhat fixes the disadvantages of this simple approach here.

timosachsenberg committed:
include first prefix ions

timosachsenberg committed:
test proforma

timosachsenberg committed:
Enhance AScore functionality with ProForma string support and comprehensive tests using TheoreticalSpectrumGenerator

timosachsenberg commented:
> I think this could be improved a lot by _actually_ calculating cumulative scores and not always the full binomial probability. Then you also would not need the weird permutation generation in the beginning (for which I am not sure if I like the bitmask version better).

It’s not a speed issue. The bitmask will probably make it easier when dealing with multiple mods in the future.

I am still not 100% sure how reassignment should handle PhosphoDecoys though. Currently they might reassign Phospho to neighboring decoy sites. Also there seem to be more permutations with same scores (because of less site determining ions if also decoys are possible).

> 
> But if this fixes things, go ahead. The stable version at least somewhat fixes the disadvantages of this simple approach here.

jpfeuffer commented:
Sorry, what's reassignment in this case? I am not too familiar with the algorithm.

timosachsenberg commented:
The current implementation annotates the highest scoring permutation. If decoys are enabled this can mean that PAPT(Phospho) becomes PA(PhosphoDecoy)PT if this leads to a higher score. More complex combinations also exist if multiple Phospho events are observed.

jpfeuffer commented:
Ah, I guess ideally one would at least report the best decoy score and the best target score but do the reassignment to the best target. But there is probably something more sound.

timosachsenberg commented:
https://pubs.acs.org/doi/full/10.1021/acs.jproteome.1c00827

timosachsenberg committed:
Enhance AScore class with detailed documentation and improve scoring logic

- Added comprehensive Doxygen-style comments to the AScore class methods, constructors, and member functions to improve code readability and maintainability.
- Updated the compute method to handle edge cases more robustly, including early termination for empty spectra and improved handling of phosphorylation site counts.
- Adjusted the scoring mechanism to ensure that the highest scoring permutations are accurately reflected in the output.
- Modified unit tests to reflect changes in scoring logic, ensuring that expected values are consistent with the new implementation.
- Updated ProForma string generation to accurately represent phosphorylation site scores.

timosachsenberg committed:
Refactor AScore and SimpleSearchEngineAlgorithm for improved PhosphoDecoy handling and add new methods for site identification

timosachsenberg committed:
Merge branch 'develop' into feature_ascore

timosachsenberg commented:
close because we do a broader implementation of rescoring algorithms in pyOpenMS. Including proper benchmarking and comparison to existing tools
